PC/_wmimodule.cpp

```diff
@@ -0,0 +1,310 @@
+            VARIANT propValue;
+            long flavor;
+            hr = value->Next(0, &propName, &propValue, NULL, &flavor);
+            if (hr == WBEM_S_NO_MORE_DATA) {
+                hr = 0;
+                break;
+            }
+            if (SUCCEEDED(hr) && (flavor & WBEM_FLAVOR_MASK_ORIGIN) != WBEM_FLAVOR_ORIGIN_SYSTEM) {
+                WCHAR propStr[8192];
+                hr = VariantToString(propValue, propStr, sizeof(propStr) / sizeof(propStr[0]));
+                if (SUCCEEDED(hr)) {
+                    DWORD cbStr1, cbStr2;
+                    DWORD written;
+                    cbStr1 = (DWORD)(wcslen(propName) * sizeof(propName[0]));
+                    cbStr2 = (DWORD)(wcslen(propStr) * sizeof(propStr[0]));
+                    if (!WriteFile(data->writePipe, propName, cbStr1, &written, NULL) ||
+                        !WriteFile(data->writePipe, (LPVOID)L"=", 2, &written, NULL) ||
+                        !WriteFile(data->writePipe, propStr, cbStr2, &written, NULL) ||
+                        !WriteFile(data->writePipe, (LPVOID)L"\0", 2, &written, NULL)
+                    ) {
+                        hr = HRESULT_FROM_WIN32(GetLastError());
+                    }
+                }
+                VariantClear(&propValue);
+                SysFreeString(propName);
+            }
+        }
+        value->EndEnumeration();
+        value->Release();
+    }
+
+    if (bstrQuery) {
+        SysFreeString(bstrQuery);
+    }
+    if (enumerator) {
+        enumerator->Release();
+    }
+    if (services) {
+        services->Release();
+    }
+    if (locator) {
+        locator->Release();
+    }
+    CoUninitialize();
+    CloseHandle(data->writePipe);
+    return (DWORD)hr;
+}
+
+
+/*[clinic input]
+_wmi.exec_query
+
+    query: unicode
+
+Runs a WMI query against the local machine.
+
+This returns a single string with 'name=value' pairs in a flat array separated
+by null characters.
+[clinic start generated code]*/
+
+static PyObject *
+_wmi_exec_query_impl(PyObject *module, PyObject *query)
+/*[clinic end generated code: output=a62303d5bb5e003f input=48d2d0a1e1a7e3c2]*/
+
+/*[clinic end generated code]*/
+{
+    PyObject *result = NULL;
+    HANDLE hThread = NULL;
+    int err = 0;
+    WCHAR buffer[8192];
+    DWORD offset = 0;
+    DWORD bytesRead;
+    struct _query_data data = {0};
+
+    if (PySys_Audit("_wmi.exec_query", "O", query) < 0) {
+        return NULL;
+    }
+
+    data.query = PyUnicode_AsWideCharString(query, NULL);
+    if (!data.query) {
+        return NULL;
+    }
+
+    if (0 != _wcsnicmp(data.query, L"select ", 7)) {
+        PyMem_Free((void *)data.query);
+        PyErr_SetString(PyExc_ValueError, "only SELECT queries are supported");
+        return NULL;
+    }
+
+    Py_BEGIN_ALLOW_THREADS
+
+    if (!CreatePipe(&data.readPipe, &data.writePipe, NULL, 0)) {
+        err = GetLastError();
+    } else {
+        hThread = CreateThread(NULL, 0, _query_thread, (LPVOID*)&data, 0, NULL);
```

Comment:
Instead of using a pipe, the worker thread could be written to modify the buffer directly. The definition of `struct _query_data` could include the query, a pointer to the result buffer, the buffer size, the size or required size of the result, and the result status. Initially use a stack allocated buffer. If the buffer is too small, have the thread keep processing without writing to the buffer in order to calculate the required size and then exit with the result status `HRESULT_FROM_WIN32(ERROR_MORE_DATA)`. In this case, allocate the buffer on the heap with the required size, and try again with a new thread. Each pass, wait on the thread with an adjusted timeout that's provided by the caller. If the request times out, kill the thread with `TerminateThread(hThread, 1)`, and fail the call.

Reply:
This is what I implemented at first, but it's much simpler to use the pipe. `ReadFile` then handles both data and cancellation (via the thread closing the pipe) directly.

I think it's possible to plumb through Ctrl+C cancellation, but I haven't figured it out yet. Probably needs synchronous overlapped IO though, which isn't too bad to do.